Added ratelimit to handle throtling cache

pierresouchay opened the pull request:
This implements a solution for https://github.com/hashicorp/consul/issues/7863

It does:
 * Add a new config `cache.entry_fetch_rate` to limit the number of calls/s for a given cache entry, default value = 1
 * Add cache.entry_fetch_max_burst size of rate limit (default value = 2)

The new configuration now supports the following syntax for instance to allow 1 query every 3s:
 * command line HCL: `-hcl 'cache = { entry_fetch_rate = 0.333}'`
 * in JSON
```json
{
  "cache": {
    "entry_fetch_rate": 0.333
  }
}
```

So, by default, the cache does not get updates faster than 1/s, even if the service is modified more frequently than that. Having such value ensures it will not modify the default behavior too much (in our case, we plan to have a value around 0.2 aka 1 modification every 5s). Having a default burst of 2 ensures that on a stable service, you get updates quickly anyway (for instance if you have a probe), so on a stable service, if you register a new service to get the latency at registration, you will see the service appear/disappear almost immediately with default settings

pierresouchay commented:
@banks @i0rek Does this mechanism to throttle cache would suit you?

dnephin commented on agent/config/runtime.go:
```diff
@@ -22,6 +22,17 @@ type RuntimeSOAConfig struct {
+}
+
+// CacheConfiguration represent the cache configuration of agent
+type CacheConfiguration struct {
+	CacheRateLimitPerEntry CacheRateLimitPerEntry
```
I think `Cache` is implied by the name of the struct.  It may also not be obvious what is being rate limited. How about something like `EntryFetchRateLimit` ?

pierresouchay replied:
ok, I'll change that

dnephin commented on agent/cache/ratelimit.go:
```diff
@@ -0,0 +1,113 @@
+// start is internal only as the ticker
+// is started by default
+func (r *rateLimiterTicker) start() {
+	go func() {
+		for {
+			select {
+			case <-r.stop:
```
I haven't looked over this implementation in detail yet, but I was surprised to see it using a ticker and a goroutine.

I think it might be possible to track counts using buckets of time, and then remove old buckets after the time has passed. I think that is the algorithm used in https://github.com/juju/ratelimit. I haven't thought too much about the trade-offs here. What do you see as the advantage to this goroutine+ticker approach?

I'm not sure if it would make sense to use https://github.com/juju/ratelimit, or maybe we could implement something smaller based on the same algorithm.

dnephin replied:
Or how about https://godoc.org/golang.org/x/time/rate#Limiter ? Would that work? I think we already have `x/time` as a dependency, so it wouldn't require adding a new one.

pierresouchay replied:
I did not know about x/time/rate and I asked @banks whether I should be using a 3rd party or implement my own, he told me to implement my own to avoid having too many dependencies, so I did :-)

I can switch to x/time/rate if you prefer, I am ok with that as it has all the required stuff in it

banks replied:
Perhaps we missunderstood each other Pierre but I certainly meant to say that we should just use `rate.Limit` just like we do everywhere else in the code 😄 . Apologies if that wasn't clear.

E.g. 
https://github.com/hashicorp/consul/blob/51c3a605ad4fc678a08901c107483203319f5949/agent/consul/config.go#L20
https://github.com/hashicorp/consul/blob/51c3a605ad4fc678a08901c107483203319f5949/agent/consul/config.go#L433

pierresouchay replied:
ok, I am revamping this then.

pierresouchay replied:
Done in https://github.com/hashicorp/consul/pull/8226/commits/bcb2fd7bff756c1535cea41d513731079c04c5a7

dnephin commented on lib/ratelimitspec.go:
```diff
@@ -0,0 +1,66 @@
+)
+
+// RateLimitSpec defines how rate limit is expressed
+type RateLimitSpec struct {
+	// burstSize defines the size of bursts for the period
+	BurstSize int
+	// Period of ticker
+	Period time.Duration
+}
```
If we do want to support multiple fields of config, I think it might be better to accept fields instead of creating a micro-format for this. But we can figure out the UX once we settle on the rate limiter implementation.

pierresouchay replied:
Burstsize, might be configurable, but I hard-coded at 2 for now. For pure cache, I don't think it deserves more configuration as I don't see how an operator might need to tune it for now (might be useful in the future - who knows)

Regarding time, see https://github.com/hashicorp/consul/issues/7863#issuecomment-631455816 => I think expressing it as "query per something" is nicer to end-user (aka: 1/s or 1/10s), but I don't care, I have been told to express it as query per something in the issue, so I did.

pierresouchay replied:
I use the same config type as RPCLImit now

pierresouchay commented on agent/cache/cache.go:
```diff
@@ -121,23 +123,29 @@ type ResultMeta struct {
+}
+
+// NewOptions build options from a RateLimitSpec
+func NewOptions(rateLimitSpec *lib.RateLimitSpec) *Options {
+	return &Options{rateLimitSpec}
 }
 
 // New creates a new cache with the given RPC client and reasonable defaults.
 // Further settings can be tweaked on the returned value.
-func New(*Options) *Cache {
+func New(options *Options) *Cache {
```
The *Options was already there, so you want me to change  `options *Options` into `options Options`?

pierresouchay replied:
Done in 48e50fe1a119b0b9480eb7c81154ff44a517d9f9

pierresouchay commented on lib/ratelimitspec.go:
```diff
@@ -0,0 +1,66 @@
+// NewRateLimitSpec builds a new RateLimitSpec with default burst of 2.
+func NewRateLimitSpec(period time.Duration) *RateLimitSpec {
+	return &RateLimitSpec{
+		2,
```
2 did make sense to me (we allow to consecutive refreshes quickly, for instance in case of a health flap, but enforce really rate limit) and wanted something simple, but yes, we can make this configurable

pierresouchay replied:
Now configurable with next patch

pierresouchay commented on agent/config/runtime_test.go:
```diff
@@ -6749,6 +6767,15 @@ func TestSanitize(t *testing.T) {
+                },
+                "Value": "10/s"
+            }
+        },
```
fixed

pierresouchay commented on agent/config/runtime.go:
```diff
@@ -443,6 +454,9 @@ type RuntimeConfig struct {
 	BootstrapExpect int
 
+	// CacheConfiguration represent cache configuration of agent
+	CacheConfiguration CacheConfiguration
```
done in next patch

pierresouchay commented on lib/ratelimitspec_test.go:
```diff
@@ -0,0 +1,60 @@
+
+	// All remaining values should be errors in parsing
+	reqError := func(toParse string) {
+		t.Helper()
```
done in https://github.com/hashicorp/consul/pull/8226/commits/4c24f1c21fa23e70498c54a955adf0a85d36b6f8

pierresouchay commented on lib/ratelimitspec.go:
```diff
@@ -0,0 +1,66 @@
+package lib
```
DONE in https://github.com/hashicorp/consul/pull/8226/commits/ec5138b11a0863f679a2495e011e7a7bc515fe82

pierresouchay commented on agent/cache/cache.go:
```diff
@@ -121,23 +123,29 @@ type ResultMeta struct {
 		entries:           make(map[string]cacheEntry),
 		entriesExpiryHeap: h,
 		stopCh:            make(chan struct{}),
+		rateLimitSpec:     options.RateLimitSpec,
```
Done

pierresouchay commented:
Hello @dnephin ,
> 
> I think it might be easier to review this as a smaller PR. What do you think about stashing most of the UX changes (config, and docs) in another branch and initially making this PR just the rate limiter and integration with the cache? That way we can focus the discussion on the core implementation, and once that is worked out we can figure out the UX.

Smaller PRs have drawbacks: we sometimes have to wait a long time before having a review and having 3 PRs instead of 1 is hard to follow on both sides (ours and yours) furthermore, we backport those changes as well on your versions and having separated versions we have sometimes to rebase make it too hard for us. And finally, it is sometimes hard to provide a feature partially implemented (what kind of split do you want ? A first with an hardcoded limit ?)

pierresouchay commented:
hello @dnephin & @i0rek,

Using the built-in cache from Golang reduces quite a lot the size of PR, so, I simplified everything, there are fewer data structures and I am using the same semantics for RPC than the current rate limiter in RPC.

I also added an end-to-end test that validates the behavior in both directions (aka to have a minimum of updates and a maximum). The test was tricky to implement because of timing issues, but I added enough tolerance to work reliably on my machine with concurrency.

dnephin commented on agent/agent_test.go:
```diff
@@ -856,6 +857,98 @@ func testAgent_AddServiceNoRemoteExec(t *testing.T, extraHCL string) {
+		test.Run(fmt.Sprintf("rate_limit_at_%v", currentTest.rateLimit), func(t *testing.T) {
+			tt := currentTest
+			t.Parallel()
+			a := NewTestAgent(t, fmt.Sprintf("cache = { rate_limit_per_entry = %v, rate_limit_burst = 1 }", tt.rateLimit))
```
Do we need a full agent to test this change?

It seems like the change is fairly localized to the `agent/cache` package, so I would expect the test to be in that same package. If we use a fake `cache.Type` implementation the test should read better as there will be fewer things that are not related to the behaviour we are trying to test. A fake implementation could report on the number of times it was called, which should make it easy to test that refresh is blocked on a limit.

That should also make it easier to show that different keys do not share a rate limit.

pierresouchay replied:
I think it is the best proof it is working from end-to-end. What would you suggest exactly?

I am not very familiar with Mocks (and to be fair, not always a big fan), but if you point me to an example of what you have in mind...

dnephin commented on agent/agent_test.go:
```diff
@@ -4201,6 +4294,8 @@ func TestAgentCache_serviceInConfigFile_initialFetchErrors_Issue6521(t *testing.
 	a2 := StartTestAgent(t, TestAgent{Name: "Agent2", HCL: `
 		server = false
 		bootstrap = false
+		// This test assumes there is not ratelimit in cache
```
Is this because the test does lots of fetches? Isn't the default value no rate limit? Why do we need to set a limit here?

pierresouchay replied:
because the timing of this test assumes all the data is immediately present in the cache. Not configuring it make this specific test fail

i0rek replied:
This shouldn't be necessary if the default is `rate.Inf`.

pierresouchay replied:
Done

dnephin commented on agent/cache/cache.go:
```diff
@@ -80,6 +81,8 @@ type Cache struct {
 	// stopCh is closed when Close is called
 	stopCh chan struct{}
+	// options includes a per Cache Rate limiter specification to avoid performing too many queries
+	options *Options
```
I do think a non-pointer `Options` would be better (similar to the constructor for the cache), but it's minor and easy enough to change later.

pierresouchay replied:
Done in 48e50fe1a119b0b9480eb7c81154ff44a517d9f9

dnephin commented on agent/cache/cache.go:
```diff
@@ -121,23 +124,26 @@ type ResultMeta struct {
 // Options are options for the Cache.
 type Options struct {
-	// Nothing currently, reserved.
+	// EntryFetchMaxBurst max burst size of RateLimit for a single cache entry
+	EntryFetchMaxBurst int
```
According to https://godoc.org/golang.org/x/time/rate#Limiter.Burst it's the number of tokens per `Wait` call, which I believe is also going to be 1. Do we need to make this configurable? Or can we remove it and always use 1 ?

pierresouchay replied:
Yes, the bust rate is important: it allows the rate-limit to be overridden for burst-size time (defaults to 2), but globally respect the rate-limit (in our case, we will probably configure it with a value around 4 or 5 I think and 0.1 qps => you can have quick updates on a stable service, but when too many updates are firing on a service, you get rate-limited at 1 query every 10s) I use the same configuration as the rate-limit you already implemented for RPC which includes burst size as well.

dnephin replied:
I think either I have misunderstood the docs I linked, or the docs are misleading, or it doesn't work quite that way. The docs suggest this only impacts `Wait()`, and we always use `Wait()` with a value of 1 , because we only ever fetch an entry one at a time. I'll need to read the code to better understand how burst works.

pierresouchay replied:
@dnephin This is exactly how it works for RPC Max Rate limit as well: https://github.com/hashicorp/consul/blob/master/agent/agent.go#L1404

dnephin replied:
Thank you both for the clarification!

It sounds like this setting is actually the bucket size. I wonder if hat would have been a better name for the config field. I guess since we already call this burst elsewhere in the config we probably have to keep it burst here as well.

pierresouchay replied:
To me, burst make sense from a functional point of view: the operator want to allow burst of data until the rate limit is reached. Bucket is more technical about the implementation. But as you highlighted it, it is anyway already used that way in RPC, so I think it make sense to change it

dnephin commented on agent/cache/cache.go:
```diff
@@ -452,7 +458,7 @@ func (c *Cache) fetch(key string, r getOptions, allowNew bool, attempt uint, ign
 	// as invalid so that it isn't returned as a valid value for a zero index.
 	if !ok {
-		entry = cacheEntry{Valid: false, Waiter: make(chan struct{})}
+		entry = cacheEntry{Valid: false, Waiter: make(chan struct{}), RateLimiter: rate.NewLimiter(c.options.EntryFetchRateLimit, c.options.EntryFetchMaxBurst)}
```
minor: now that this is really long it might read better with newlines for each field:

```suggestion
		entry = cacheEntry{
			Valid: false,
			Waiter: make(chan struct{}),
			RateLimiter: rate.NewLimiter(c.options.EntryFetchRateLimit, 1),
		}
```

pierresouchay replied:
Fixed in https://github.com/hashicorp/consul/pull/8226/commits/1e0b1d2886f2108d18f4385d7ed1625ae6a946ec

dnephin commented on agent/cache/entry.go:
```diff
@@ -41,6 +43,8 @@ type cacheEntry struct {
 	// background request has be blocking for at least 5 seconds, which ever
 	// happens first.
 	RefreshLostContact time.Time
+	// Ratelimit avoid entries to fetch too often too many results
+	RateLimiter *rate.Limiter
```
```suggestion
	// FetchRatelimiter limits the rate at which fetch is called for this entry.
	FetchRatelimiter *rate.Limiter
```

dnephin commented on agent/cache/cache.go:
```diff
@@ -499,7 +505,13 @@ func (c *Cache) fetch(key string, r getOptions, allowNew bool, attempt uint, ign
 			}
 		}
-
+		if err := entry.RateLimiter.Wait(context.Background()); err != nil {
+			// This should might happen only if:
+			// burstSize is 0, Context is canceled, or the expected wait time exceeds the Context's Deadline.
+			// so, it should not happen, fix this if you plan using something different than:
+			// context.Background()
```
Just to be extra sure we don't miss it, I think a comment above the line about the context would be good

```suggestion
		// TODO(context): handle the error without a panic when context is changed to one that is cancellable
		if err := entry.RateLimiter.Wait(context.Background()); err != nil {
			// an error here is a programming error (bursSize is 0) so it is ok to panic, until the context
			// is changed to one that can be cancelled
```

dnephin commented on agent/config/config.go:
```diff
@@ -58,6 +58,14 @@ func Parse(data string, format string) (c Config, md mapstructure.Metadata, err
 
+// CacheConfig Cache tunning configuration
+type CacheConfig struct {
+	// EntryFetchMaxBurst max burst size of RateLimit for a single cache entry
+	EntryFetchMaxBurst *int `json:"rate_limit_burst,omitempty" hcl:"rate_limit_burst" mapstructure:"rate_limit_burst"`
```
Similar to above, I think we can remove burst, because it should always be 1 (unless I have misunderstood something).

dnephin commented on agent/config/config.go:
```diff
@@ -58,6 +58,14 @@ func Parse(data string, format string) (c Config, md mapstructure.Metadata, err
 	return c, md, nil
 }
 
+// CacheConfig Cache tunning configuration
+type CacheConfig struct {
```
```suggestion
// Cache tunning configuration
type Cache struct {
```

dnephin commented on agent/config/config.go:
```diff
@@ -101,6 +109,7 @@ type Config struct {
 	BindAddr                         *string                  `json:"bind_addr,omitempty" hcl:"bind_addr" mapstructure:"bind_addr"`
 	Bootstrap                        *bool                    `json:"bootstrap,omitempty" hcl:"bootstrap" mapstructure:"bootstrap"`
 	BootstrapExpect                  *int                     `json:"bootstrap_expect,omitempty" hcl:"bootstrap_expect" mapstructure:"bootstrap_expect"`
+	CacheConfig                      CacheConfig              `json:"cache,omitempty" hcl:"cache" mapstructure:"cache"`
```
```suggestion
	Cache                      Cache              `json:"cache,omitempty" hcl:"cache" mapstructure:"cache"`
```

(although this change will need a `gofmt`)

pierresouchay replied:
This is in conflict with Cache defined in configuration:
```
agent/config/runtime.go:26:6: Cache redeclared in this block
	previous declaration at agent/config/config.go:62:6
```
(the one defined in agent/config/config.go has optional fields EntryFetchMaxBurst/EntryFetchRateLimit while those are mandatory for CacheConfig).

I could re-use the same structure, but in that case EntryFetchMaxBurst and EntryFetchRateLimit would be pointers, not values... it is what you want?

dnephin replied:
Ah, right! I forgot there are these 2 structs in `config`. I wonder if we can use `agent/cache.Options` in `RuntimeConfig` , and use `Cache` in `Config`. That way we don't need to translate it again. We can use the value directly from `RuntimeConfig` in  `cache.New()`

pierresouchay replied:
Done in 48e50fe1a119b0b9480eb7c81154ff44a517d9f9

dnephin commented on agent/config/runtime.go:
```diff
@@ -22,6 +22,14 @@ type RuntimeSOAConfig struct {
 
+// Cache represent the cache configuration of agent
+type Cache struct {
+	// EntryFetchMaxBurst max burst size of RateLimit for a single cache entry
+	EntryFetchMaxBurst int
```
same as above, I think we can remove

pierresouchay replied:
What do you think about my comment https://github.com/hashicorp/consul/pull/8226#discussion_r459112485 ?

dnephin commented on agent/config/config.go:
```diff
@@ -58,6 +58,14 @@ func Parse(data string, format string) (c Config, md mapstructure.Metadata, err
+	// EntryFetchMaxBurst max burst size of RateLimit for a single cache entry
+	EntryFetchMaxBurst *int `json:"rate_limit_burst,omitempty" hcl:"rate_limit_burst" mapstructure:"rate_limit_burst"`
+	// EntryFetchRateLimit represents the max calls/sec for a single cache entry
+	EntryFetchRateLimit *float64 `json:"rate_limit_per_entry,omitempty" hcl:"rate_limit_per_entry" mapstructure:"rate_limit_per_entry"`
```
```suggestion
	EntryFetchRateLimit *float64 `mapstructure:"entry_fetch_rate_limit"`
```

pierresouchay replied:
fixed in https://github.com/hashicorp/consul/pull/8226/commits/1e0b1d2886f2108d18f4385d7ed1625ae6a946ec

dnephin commented on website/pages/docs/agent/options.mdx:
```diff
@@ -955,6 +955,22 @@ Valid time units are 'ns', 'us' (or 'µs'), 'ms', 's', 'm', 'h'."
+    for a long time, 2 successive queries can be made as long as the rate-limit is not
+    reached.
+
+  - `rate_limit_per_entry`: configures the rate-limit at which the cache may refresh a single
```
```suggestion
  - `entry_fetch_rate_limit`: configures the rate-limit at which the cache may refresh a single
```

pierresouchay replied:
ok, I'll change that tomorrow

pierresouchay replied:
Fixed in https://github.com/hashicorp/consul/pull/8226/commits/00df40b9f5a5f15d0e55095c06b0a007bd74db6e

dnephin commented on agent/config/builder.go:
```diff
@@ -882,11 +882,15 @@ func (b *Builder) Build() (rt RuntimeConfig, err error) {
+		Bootstrap:        b.boolVal(c.Bootstrap),
+		BootstrapExpect:  b.intVal(c.BootstrapExpect),
+		Cache: Cache{
+			EntryFetchRateLimit: rate.Limit(b.float64ValWithDefault(c.CacheConfig.EntryFetchRateLimit, 1)),
```
The default should be `rate.Inf`, no?

pierresouchay replied:
Well, that's debatable :-)

As explained in the header of this PR (and in the doc), I have chosen that's the default value is 1qps by default with a burst size of 2.
It means the cache will not do more than 1 qps on servers by default. It sounds reasonable to have such value because otherwise, the cache is doing more harm than dumb calls from RPC APIs.

Of course, that's to HashiCorp to choose the value, but having large values there will break your clients as soon as they use cache. Using it is easy: use Consul Connect. In our experience, greater values than that just hurts the cluster while bringing no benefit. Let this value to Inf is IMHO a mistake.

As proof: all integration tests work with this (including all connect tests), so I think having it a 1qps by default is a good choice, but of course, you choose the value your want.

dnephin replied:
I think it needs to be effectively disabled by default, at least for the initial release. My understanding is that using `rate.Inf` will be the equivalent to disabled. If we were to enable it I think that would be a breaking change. Other users may have smaller clusters and might expect a lot more than 1qps.

In a future major release we could choose to enable it with a different value.

pierresouchay replied:
Ok, you are the boss, I'll change that tomorrow but I think it is a mistake: the goal of a cache is to improve performance of reads, not to break the cluster :-)

Having sub-second micro-updates for 1h in the cache is not gonna make Consul more efficient, it will break Consul Connect and force people to tune it *AFTER* an outage :-)

dnephin replied:
I think we can probably set a default value here in the future, but I think we want this to go into 1.8.x, so I think we have to be a little more conservative. In 1.9 we could set this to something smaller.

pierresouchay replied:
@dnephin 3 qps by default? Much more?

pierresouchay replied:
@dnephin I did put 2 qps instead of 1 by default in https://github.com/hashicorp/consul/pull/8226/commits/593b18b199f87c099bde316d7f11bad9376a4aa8 and now, you just have to change a constant to change the default values (tests are not impacted anymore).

Feel free to tell me the exact value you want (after discussion with your team I imagine)

i0rek replied:
For 1.8.x we should set it to `rate.Inf` and have a discussion if we want to change it for 1.9.

pierresouchay replied:
Done in b3947c3f218b1e616c05d8f4cb0876cf7a543cd6

pierresouchay replied:
Set to +Inf

dnephin commented:
> Smaller PRs have drawbacks

There are always trade-offs! Larger PRs can take longer to review and require more back and forth, but it is true that if you have to wait a while for an initial review, maybe it is easier to do more at once. Hopefully the ci automation makes it easier to split things up a bit more and get them all cherry-picked into the release branches.

> it is sometimes hard to provide a feature partially implemented 

That is very true. It's hard to say exactly where the right line is. Sometime what I try to do is open one small PR with the core part of the change, and then a second (or sometimes more) draft PRs that help demonstrate how that core feature might be exposed to the user, or integrated in other places. The draft PR doesn't need to be totally ready for review, so it can save some time up front. But maybe that approach won't always work.

> In any case, I am trying to provide PRs with clear separation in commits that are atomic, do work, compile, and pass tests.

Thank you! We really appreciate your PRs!

i0rek commented on agent/cache/cache.go:
```diff
@@ -499,7 +512,12 @@ func (c *Cache) fetch(key string, r getOptions, allowNew bool, attempt uint, ign
 		}
-
+		// TODO(context): handle the error without a panic when context is changed to one that is cancellable
+		if err := entry.FetchRateLimiter.Wait(context.Background()); err != nil {
```
We have `lib.StopChannelContext{StopCh: shutdownCh}` that should be used here to cancel when shutting down. I think this is the least we should be doing here. I don't think the cache has access to the agent shutdown channel already though.

pierresouchay replied:
Fixed in 7b23877ec74d40aca69941f890188ba4ae11dc91

i0rek commented on agent/config/builder.go:
```diff
@@ -32,6 +33,15 @@ import (
+// Ensure to update documentation if you modify those values
+const (
+	// DefaultEntryFetchMaxBurst is the default value for cache.entry_fetch_max_burst
+	DefaultEntryFetchMaxBurst = 2
```
The rate limiter should be disabled by default imho. Which means that the default should be `rate.Inf`.

pierresouchay replied:
Done in b3947c3f218b1e616c05d8f4cb0876cf7a543cd6

i0rek commented on agent/config/config.go:
```diff
@@ -58,6 +58,14 @@ func Parse(data string, format string) (c Config, md mapstructure.Metadata, err
+	// EntryFetchMaxBurst max burst size of RateLimit for a single cache entry
+	EntryFetchMaxBurst *int `json:"entry_fetch_max_burst,omitempty" hcl:"entry_fetch_max_burst" mapstructure:"entry_fetch_max_burst"`
+	// EntryFetchRateLimit represents the max calls/sec for a single cache entry
+	EntryFetchRateLimit *float64 `json:"entry_fetch_rate_limit,omitempty" hcl:"entry_fetch_rate_limit" mapstructure:"entry_fetch_rate_limit"`
```
The rpc equivalent is called `rpc_rate`, what do you think about using `entry_fetch_rate` here?

pierresouchay replied:
Done in b3947c3f218b1e616c05d8f4cb0876cf7a543cd6

i0rek commented on agent/config/config.go:
```diff
@@ -101,6 +109,7 @@ type Config struct {
 	BindAddr                         *string                  `json:"bind_addr,omitempty" hcl:"bind_addr" mapstructure:"bind_addr"`
 	Bootstrap                        *bool                    `json:"bootstrap,omitempty" hcl:"bootstrap" mapstructure:"bootstrap"`
 	BootstrapExpect                  *int                     `json:"bootstrap_expect,omitempty" hcl:"bootstrap_expect" mapstructure:"bootstrap_expect"`
+	Cache                            Cache                    `json:"cache,omitempty" hcl:"cache" mapstructure:"cache"`
```
The rpc equivalent is in `limits`. Since there is no existing `cache` config just yet, I am wondering if putting it into `limits` is better. If so I would lean towards naming the options `cache_limit` and `cache_max_burst`.

pierresouchay replied:
I don't know, I personally like to have structured data (and I initially planned to have more complex cache configuration), hence the dedicated section, but this could be changed. Tell me if you want me to change it.

@dnephin do you have an opinion?

pierresouchay replied:
@i0rek There is a downside to move it into limits: {}

=> now cache {} is consistent between Runtime and Configuration (aka same name & so on).
@dnephin asked me to use `cache.Config` in the runtime (so we do not duplicate and use directly in constructor of Cache)

So we have:

Config.Cache.entry_fetch* to configure
And
Runtime.Cache.entry* to see applied configuration (this is consistent, and shown properly in /v1/agent/self


If I move to limits, I'll still have to keep Cache in Runtime because it is actually a `cache.Options` as requested by @dnephin 

The good thing about cache is Consistency (which is cool for a cache 🥇 ) between Runtime and Config.
Furthermore, limits are currently dispatched at various places in Runtime as flat... (I personally hate that Runtime as shown in /v1/agent/self and config have completely different layouts and names, this is hard to compare actual values in server and config, but there are probably too many people relying on those names)

dnephin replied:
I think a cache section makes more sense. I like the idea of grouping by components or sub-systems, I think that is a more natural way to organize them.

In the future we might want to make it `api.cache`, but I think that would be better as a larger change to more config fields.

pierresouchay replied:
@dnephin I feel the same (and would like to avoid changing it one more time :-) - But if you decide otherwise, I'll do it

i0rek commented on website/pages/docs/agent/options.mdx:
```diff
@@ -955,6 +955,24 @@ Valid time units are 'ns', 'us' (or 'µs'), 'ms', 's', 'm', 'h'."
+- `cache` Cache configuration of agent. The configurable values are the following:
+
+  - `entry_fetch_max_burst`: The size of the token bucket used to recharge the rate-limit per
+    cache entry. The default value is 2 and means that when cache has not been updated
```
The default is now `rate.Inf`.

pierresouchay replied:
@i0rek The `cache.entry_fetch_max_burst` is actually 2 by default (even if it has no real meaning until `cache.entry_fetch_rate` is set to something different than +Inf...
I propose to add: The default value is 2 if `cache.entry_fetch_rate` is set.

i0rek replied:
I am sorry, I looked at the wrong thing!

ltagliamonte-dd commented:
@i0rek @dnephin would be great to have a 1.7.X release with this change as well. Thanks
